File: experiments/mspti_sync_skeleton/test_buffer_audit.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import tempfile
from pathlib import Path

from buffer_audit import aggregate_buffer_audits, validate_rank_buffer_audit


def _payload(rank: int = 0) -> dict:
    counters = {
        "buffer_request_count": 2,
        "buffer_complete_callback_count": 2,
        "buffer_complete_count": 2,
        "inflight": 0,
        "inflight_final": 0,
        "max_inflight": 2,
        "parsed_kernel": 3,
        "parsed_comm": 1,
        "parsed_unknown": 0,
        "enqueued_kernel": 3,
        "enqueued_comm": 1,
        "queue_rejected_kernel": 0,
        "queue_rejected_comm": 0,
        "queue_rejected_other": 0,
        "queue_rejected_total": 0,
        "worker_consumed_kernel": 3,
        "worker_consumed_comm": 1,
        "processed_kernel": 3,
        "processed_comm": 1,
        "emitted_total": 4,
        "emitted_comm": 1,
        "emitted_kseg": 1,
        "emitted_kseg_raw_count": 3,
        "valid_while_stopping": 0,
        "callback_wall_total_ns": 40,
        "callback_wall_max_ns": 30,
        "malformed_timestamp_count": 0,
        "malformed_sample_count": 0,
        "ledger_capacity": 4096,
        "ledger_entry_count": 2,
        "ledger_completed_count": 2,
        "ledger_overflow_count": 0,
    }
    conservation = {
        "request_complete_balanced": True,
        "kernel_parse_queue_balanced": True,
        "kernel_worker_balanced": True,
        "kernel_process_balanced": True,
        "kernel_emit_balanced": True,
        "comm_parse_queue_balanced": True,
        "comm_worker_balanced": True,
        "comm_process_balanced": True,
        "comm_emit_balanced": True,
        "unknown_kind_zero": True,
        "valid_while_stopping_zero": True,
        "inflight_final_zero": True,
        "ledger_complete_balanced": True,
        "ledger_no_overflow": True,
        "fingerprint_all_ok": True,
        "lifecycle_snapshots_complete": True,
        "lifecycle_ordered": True,
        "before_free_quiescent": True,
        "all_ok": True,
    }
    fingerprint = {
        "kernel": {"count": 3, "xor64": 123, "sum64": 456},
        "communication": {"count": 1, "xor64": 789, "sum64": 789},
    }
    fingerprints = {
        stage: json.loads(json.dumps(fingerprint))
        for stage in ("parsed", "enqueued", "worker", "processed", "emitted")
    }
    lifecycle_snapshots = {
        name: {
            "captured": True,
            "monotonic_ns": timestamp,
            "inflight": 0,
            "active_calls": 0,
            "worker_busy": False,
        }
        for name, timestamp in (
            ("before_disable", 100),
            ("after_flush", 200),
            ("before_free", 300),
        )
    }
    buffer_ledger = [
        {
            "request_id": 1,
            "pointer_token": 1001,
            "reuse_generation": 1,
            "requested_size": 8192,
            "completion_size": 8192,
            "valid_size": 4096,
            "callback_wall_ns": 10,
            "kernel_records": 3,
            "comm_records": 0,
            "unknown_records": 0,
            "get_next_calls": 4,
            "completed": True,
        },
        {
            "request_id": 2,
            "pointer_token": 1001,
            "reuse_generation": 2,
            "requested_size": 8192,
            "completion_size": 8192,
            "valid_size": 2048,
            "callback_wall_ns": 30,
            "kernel_records": 0,
            "comm_records": 1,
            "unknown_records": 0,
            "get_next_calls": 2,
            "completed": True,
        },
    ]
    return {
        "schema_version": 2,
        "instrumentation": "mspti_buffer_audit_v2",
        "rank": rank,
        "counters": counters,
        "conservation": conservation,
        "activity_integrity_ok": True,
        "malformed_timestamp_samples": [],
        "fingerprints": fingerprints,
        "lifecycle_snapshots": lifecycle_snapshots,
        "buffer_ledger": buffer_ledger,
    }


def test_validate_and_exact_inventory() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        root = Path(tmp)
        path = root / "rank_0000.buffer_audit.json"
        path.write_text(json.dumps(_payload()), encoding="utf-8")
        aggregate, errors = aggregate_buffer_audits(root, expected_paths={0: path})
        assert not errors
        assert aggregate["pass"] is True
        assert (root / "buffer_audit.json").is_file()


def test_missing_selected_fails_closed() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        root = Path(tmp)
        _, errors = aggregate_buffer_audits(
            root, expected_paths={0: root / "rank_0000.buffer_audit.json"}
        )
        assert any("missing_selected=[0]" in error for error in errors)


def test_unselected_and_conservation_failure() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        root = Path(tmp)
        selected = root / "rank_0000.buffer_audit.json"
        selected.write_text(json.dumps(_payload()), encoding="utf-8")
        unexpected = root / "rank_0001.buffer_audit.json"
        bad = _payload(rank=1)
        bad["counters"]["worker_consumed_comm"] = 0
        unexpected.write_text(json.dumps(bad), encoding="utf-8")
        _, errors = aggregate_buffer_audits(root, expected_paths={0: selected})
        assert any("unexpected=" in error for error in errors)
        assert any("comm_worker_balanced" in error for error in errors)


def test_malformed_timestamp_fails_integrity() -> None:
    payload = _payload()
    payload["counters"]["malformed_timestamp_count"] = 1
    payload["counters"]["malformed_sample_count"] = 1
    payload["activity_integrity_ok"] = False
    payload["malformed_timestamp_samples"] = [{"start_ns": 20, "end_ns": 10}]
    errors = validate_rank_buffer_audit(payload, expected_rank=0)
    assert any("malformed_timestamp_count=1" in error for error in errors)


def test_fingerprint_ledger_and_lifecycle_corruption_fail() -> None:
    payload = _payload()
    payload["fingerprints"]["processed"]["communication"]["xor64"] += 1
    payload["conservation"]["fingerprint_all_ok"] = False
    payload["conservation"]["all_ok"] = False
    payload["activity_integrity_ok"] = False
    errors = validate_rank_buffer_audit(payload, expected_rank=0)
    assert any("fingerprint" in error for error in errors)

    payload = _payload()
    payload["buffer_ledger"][1]["reuse_generation"] = 3
    payload["lifecycle_snapshots"]["before_free"]["inflight"] = 1
    errors = validate_rank_buffer_audit(payload, expected_rank=0)
    assert any("reuse generation" in error for error in errors)
    assert any("before_free snapshot not quiescent" in error for error in errors)


if __name__ == "__main__":
    test_validate_and_exact_inventory()
    test_missing_selected_fails_closed()
    test_unselected_and_conservation_failure()
    test_malformed_timestamp_fails_integrity()
    test_fingerprint_ledger_and_lifecycle_corruption_fail()

File: experiments/mspti_sync_skeleton/test_buffer_audit.cpp
#include "buffer_audit.hpp"

#include <cassert>
#include <string>

using mspti_skeleton::AuditRawKind;
using mspti_skeleton::AuditContentStage;
using mspti_skeleton::AuditLifecyclePoint;
using mspti_skeleton::BufferAuditCounters;
using mspti_skeleton::BufferAuditPathForOutput;
using mspti_skeleton::CheckBufferAuditConservation;
using mspti_skeleton::MalformedTimestampSample;

int main() {
    BufferAuditCounters counters;
    int first_buffer = 0;
    int second_buffer = 0;
    const auto first = counters.BeginBufferRequest(
        &first_buffer, 1, 8192, 100, 1);
    const auto second = counters.BeginBufferRequest(
        &second_buffer, 1, 8192, 110, 2);
    counters.NoteBufferCompleteCallback();
    counters.NoteBufferCompleteCallback();
    assert(counters.NoteOwnedBufferComplete(first, 8192, 4096, 200) == 1);
    assert(counters.NoteOwnedBufferComplete(second, 8192, 2048, 220) == 2);
    counters.NoteLedgerTerminal(first, 7, 4);
    counters.NoteLedgerTerminal(second, 7, 2);

    for (int i = 0; i < 3; ++i) {
        const uint64_t hash = 100 + static_cast<uint64_t>(i);
        counters.NoteParsed(AuditRawKind::kKernel);
        counters.NoteWorkerConsumed(AuditRawKind::kKernel);
        counters.NoteProcessed(AuditRawKind::kKernel);
        for (uint32_t stage = 0;
             stage < static_cast<uint32_t>(AuditContentStage::kCount); ++stage) {
            counters.NoteContent(static_cast<AuditContentStage>(stage),
                                 AuditRawKind::kKernel, hash);
        }
    }
    counters.NoteEmittedKseg(3);
    counters.NoteParsed(AuditRawKind::kCommunication);
    counters.NoteWorkerConsumed(AuditRawKind::kCommunication);
    counters.NoteProcessed(AuditRawKind::kCommunication);
    for (uint32_t stage = 0;
         stage < static_cast<uint32_t>(AuditContentStage::kCount); ++stage) {
        counters.NoteContent(static_cast<AuditContentStage>(stage),
                             AuditRawKind::kCommunication, 200);
    }
    counters.NoteEmittedComm();
    counters.NoteEmittedTotal();
    counters.NoteCallbackWall(first, 10);
    counters.NoteCallbackWall(second, 30);
    counters.NoteLifecycleSnapshot(
        AuditLifecyclePoint::kBeforeDisable, 300, 0, 0, false);
    counters.NoteLifecycleSnapshot(
        AuditLifecyclePoint::kAfterFlush, 310, 0, 0, false);
    counters.NoteLifecycleSnapshot(
        AuditLifecyclePoint::kBeforeFree, 320, 0, 0, false);

    auto snapshot = counters.Snapshot();
    assert(snapshot.max_inflight == 2);
    assert(snapshot.callback_wall_total_ns == 40);
    assert(snapshot.callback_wall_max_ns == 30);
    assert(snapshot.ledger_entry_count == 2);
    assert(snapshot.ledger_completed_count == 2);
    assert(snapshot.buffer_ledger[0].valid_size == 4096);
    assert(snapshot.buffer_ledger[1].callback_wall_ns == 30);
    assert(snapshot.lifecycle_snapshots[2].captured);
    auto conservation = CheckBufferAuditConservation(
        snapshot, /*enqueued_kernel=*/3, /*enqueued_comm=*/1,
        /*inflight_final=*/0);
    assert(conservation.all_ok);

    counters.NoteParsed(AuditRawKind::kKernel);
    counters.NoteContent(AuditContentStage::kParsed, AuditRawKind::kKernel, 999);
    counters.NoteQueueRejected(AuditRawKind::kKernel);
    snapshot = counters.Snapshot();
    conservation = CheckBufferAuditConservation(
        snapshot, /*enqueued_kernel=*/3, /*enqueued_comm=*/1,
        /*inflight_final=*/0);
    assert(conservation.kernel_parse_queue_balanced);
    assert(!conservation.fingerprint_all_ok);
    assert(!conservation.all_ok);

    counters.NoteParsedUnknown();
    snapshot = counters.Snapshot();
    conservation = CheckBufferAuditConservation(snapshot, 3, 1, 0);
    assert(!conservation.unknown_kind_zero);
    assert(!conservation.all_ok);

    MalformedTimestampSample sample;
    sample.raw_kind = static_cast<uint32_t>(AuditRawKind::kCommunication);
    sample.start_ns = 20;
    sample.end_ns = 10;
    counters.NoteMalformedTimestamp(sample);
    snapshot = counters.Snapshot();
    assert(snapshot.malformed_timestamp_count == 1);
    assert(snapshot.malformed_sample_count == 1);
    assert(snapshot.malformed_samples[0].start_ns == 20);

    assert(BufferAuditPathForOutput("/tmp/rank_0007.skeleton.jsonl") ==
           "/tmp/rank_0007.buffer_audit.json");
    assert(BufferAuditPathForOutput("/tmp/custom.out") ==
           "/tmp/custom.out.buffer_audit.json");

    counters.Reset();
    snapshot = counters.Snapshot();
    assert(snapshot.buffer_request_count == 0);
    assert(snapshot.malformed_timestamp_count == 0);
    return 0;
}
